src/sentry/web/frontend/accounts.py

```diff
@@ -463,71 +469,80 @@ def show_emails(request):
 
         return HttpResponseRedirect(request.path)
 
-    if email_form.is_valid():
-        old_email = user.email
+    if 'primary' in request.POST:
+        new_primary = request.POST.get('new_primary_email').lower()
+
+        if User.objects.filter(username__iexact=new_primary).exclude(id=user.id).exists()\
+                or User.objects.filter(email__iexact=new_primary).exclude(id=user.id).exists():
+            messages.add_message(request,
+                messages.ERROR,
+                _("That email is already in use for another user")
+            )
 
-        email_form.save()
+        elif new_primary != user.email:
 
-        if user.email != old_email:
-            useroptions = UserOption.objects.filter(user=user, value=old_email)
-            for option in useroptions:
-                option.value = user.email
+            # update notification settings for those set to primary email with new primary email
+            alert_email = UserOption.objects.get_value(user=user, project=None, key='alert_email')
+
+            if alert_email == user.email:
+                UserOption.objects.set_value(user=user, project=None, key='alert_email', value=new_primary)
+            options = UserOption.objects.filter(user=user, key='mail:email')
+            for option in options:
+                if option.value != user.email:
+                    continue
+                option.value = new_primary
                 option.save()
-            UserEmail.objects.filter(user=user, email=old_email).delete()
+
+            new_username = user.email == user.username
```

Comment:
it might be nice to call this something like `has_new_username` to make it more obvious it's a boolean and not an actual username